sys/net/gnrc/network_layer/sixlowpan/frag/sfr/gnrc_sixlowpan_frag_sfr.c

```diff
@@ -0,0 +1,1179 @@
+    }
+    else {
+        DEBUG("6lo sfr: Unknown dispatch: %02x\n",
+              hdr->disp_ecn & SIXLOWPAN_SFR_DISP_MASK);
+        gnrc_pktbuf_release(pkt);
+    }
+}
+
+int gnrc_sixlowpan_frag_sfr_forward(gnrc_pktsnip_t *pkt,
+                                    sixlowpan_sfr_rfrag_t *rfrag,
+                                    gnrc_sixlowpan_frag_vrb_t *vrbe,
+                                    unsigned page)
+{
+    _generic_rb_entry_t entry = { .type = _VRB, .entry = { .vrb = vrbe } };
+    gnrc_pktsnip_t *hdrsnip = gnrc_pktbuf_add(pkt, rfrag, sizeof(*rfrag),
+                                              GNRC_NETTYPE_SIXLOWPAN);
+
+    if (hdrsnip == NULL) {
+        DEBUG("6lo sfr: Unable to allocate new rfrag header\n");
+        gnrc_pktbuf_release(pkt);
+        return -ENOMEM;
+    }
+    /* due to compression, packet length of the original fragment might have
+     * changed */
+    sixlowpan_sfr_rfrag_set_frag_size(hdrsnip->data, gnrc_pkt_len(pkt));
+    /* offset is adapted in `_forward_rfrag()` */
+    return _forward_rfrag(hdrsnip, &entry,
+                          sixlowpan_sfr_rfrag_get_offset(rfrag), page);
+}
+
+void gnrc_sixlowpan_frag_sfr_arq_timeout(gnrc_sixlowpan_frag_fb_t *fbuf)
+{
+    uint32_t now = xtimer_now_usec();
+    _window_frags_t * const head = (_window_frags_t *)fbuf->sfr.window.next;
+    _window_frags_t *frag_desc = head;
+    const uint32_t last_arq_timeout = now - (fbuf->sfr.arq_timeout * US_PER_MS);
+    uint32_t next_arq_offset = fbuf->sfr.arq_timeout;
+
+    /* copying clist_foreach because we can't work just in function context*/
+    if (frag_desc) {
+        do {
+            uint32_t diff;
+
+            frag_desc = (_window_frags_t *)frag_desc->super.next;
+            diff = frag_desc->last_sent - last_arq_timeout;
+            if (diff < (fbuf->sfr.arq_timeout * US_PER_MS)) {
+                /* this fragment's last was last sent < fbuf->sfr.arq_timeout
+                 * ago */
+                uint32_t offset = fbuf->sfr.arq_timeout - (diff / US_PER_MS);
+                if (offset < next_arq_offset) {
+                    /* wait for this fragments ACK next */
+                    next_arq_offset = offset;
+                }
+            }
+            else if (_frag_ack_req(frag_desc)) {
+                if ((frag_desc->retries++) < GNRC_SIXLOWPAN_SFR_FRAG_RETRIES) {
+                    bf_set(fbuf->sfr.to_resend.bf, _frag_seq(frag_desc));
+                    _sched_next_frame(fbuf);
+                }
+                else {
+                    DEBUG("6lo sfr: resending fragment (tag: %u, X: %i, seq: %u, "
+                          "frag_size: %u, offset: %u), giving up sending datagram\n",
+                          (uint8_t)fbuf->tag, _frag_ack_req(frag_desc),
+                          _frag_seq(frag_desc), _frag_size(frag_desc),
+                          frag_desc->offset);
+                    gnrc_pktbuf_release(fbuf->pkt);
+                    fbuf->pkt = NULL;
```

Comment:
Note to self: this leads to a duplicate release. fbuf is cleaned at error label.